ehrql/serializer.py

```diff
@@ -0,0 +1,265 @@
+import datetime
+import functools
+import json
+import pathlib
+
+from ehrql.codes import BaseCode
+from ehrql.file_formats.base import BaseDatasetReader
+from ehrql.measures.measures import Measure
+from ehrql.query_language import DummyDataConfig
+from ehrql.query_model.column_specs import ColumnSpec
+from ehrql.query_model.nodes import Node, Position, Value
+from ehrql.query_model.table_schema import BaseConstraint, Column, TableSchema
+from ehrql.utils.module_utils import get_all_subclasses
+
+
+# Mapping of names to types for all types that we support serializing/deserializing
+TYPE_REGISTRY = {
+    type_.__qualname__: type_
+    for type_ in [
+        bool,
+        int,
+        float,
+        str,
+        datetime.date,
+        tuple,
+        dict,
+        frozenset,
+        Position,
+        TableSchema,
+        Column,
+        pathlib.Path,
+        ColumnSpec,
+        DummyDataConfig,
+        Measure,
+        *get_all_subclasses(Node),
+        *get_all_subclasses(BaseCode),
+        *get_all_subclasses(BaseConstraint),
+        *get_all_subclasses(BaseDatasetReader),
```

Comment:
I don't understand why we'd need to serialize readers.

Reply:
This is part of the `table_from_file` feature. This ends up building `InlinePatientTable` instances whose `rows` attributes are `DatasetReader` instances, and we need to serialize these.